### DISCUSSION TITLE
How can I find the right Pin Mapping for the STM32F401

### DISCUSSION THREAD

**Discussion by Helle62** [grblHAL/STM32F4xx#119] [Q&A]
Hello,
bevor excuse me for my bad english and I hope my post is in the right category!?

I build different BIN-files with the Web-Builder, but I can't find the right pin mapping. I found the video from GrblGru, but the pinout is for a STM32F103 and different to my STM32F401. (Steps and Directions are reversed)

**Please can you tell me what is the best way to find out the right pin mapping?** I read the files config.h and generic_map.h, but it is too much coding for me.

The Web Builder is a great tool, with very much variations. It would be so helpful, if the Web Builder can show for the different variations a little table with the associated pin mapping.

For your help, many thanks in advance,
Helmut

**Comment by terjeio** [comment]
> I build different BIN-files with the Web-Builder, but I can't find the right pin mapping.

None of the available pin maps matches the board you have? If so a new map has to be created.

> It would be so helpful, if the Web Builder can show for the different variations a little table with the associated pin mapping.

I have just added a new button _Board map_ to the [Web Builder](http://svn.io-engineering.com:8080/) that can be used to view the map file for the selected board.
Adding the board map files in a tabulated format is not something I will do - a lot of work and hard to keep in sync with the actual map files.
However, if someone can come up with (javascript?) code to parse map files and present the mapping tabulated I am all ears...

**Comment by Helle62** [comment]
Many thanks for the fast answer!

> None of the available pin maps matches the board you have? If so a new map has to be created.

No, I found only one pin mapping from GrblGru. And I thought that I have to find the right BIN-file for this pinout, to wire my STM32F4.  But the drivers are different,  so this pin mapping does not work for me.

> However, if someone can come up with (javascript?) code to parse map files and present the mapping tabulated I am all ears...

This sounds good and would be very helpful!

I use the STM32F401 without board. I will use TB6600 driver for the steper motors. Also I have "only" to know what I have to select at Web Builder to create a BIN-file for a 3 axis machine and ganged y-axis. Plus the right pin mapping, that I know wich pin is for steps, directions, limits, ...aso.

If you say it's better to use the STM32F103 or it's easier to build the project with  Arduino Due, I will do that.

Have a nice rest of evening,
Helle

**Reply by terjeio** [reply]
> No, I found only one pin mapping from GrblGru. And I thought that I have to find the right BIN-file for this pinout, to wire my STM32F4.

You do not - GrblGru, or any other senders that I am aware of, does not care about pin mappings. You need a map that assigns pins that are available.

> I use the STM32F401 without board.

Then select a board for a F401 variant that matches the one you have. You may have to test if it works as variants may have different crystal frequencies, if not matching the firmware timing will be off and USB and/or serial communication will not work. FYI most boards use a 8MHz crystal. Worst case you will have to create a map file and build using Eclipse or platformio.
Note that the F401 map for the Nucleo-64 dev board is special, it can only be used with Nucleo-64.

> If you say it's better to use the STM32F103 or it's easier to build the project with Arduino Due, I will do that.

Not neccesarily easier if you do not have a matching CNC breakout board. A plus for the F401 is it is more capable processor and IMO better suited for grblHAL as it has a floating point unit (FPU).

**Comment by Helle62** [comment]
Many thanks again for your answer!

> FYI most boards use a 8MHz crystal. Worst case you will have to create a map file and build using Eclipse or platformio.

I can’t work with Eclipse or PlatformIO. Im worked a long time ago with Python and actually with Arduino IDE. I'm looking for a ready/tested solution. 
But I have seen today the first time the button “Board Map” in the Web Builder and I hope this shows me the right “PIN assignment” what I need.
Have a nice evening,
Helmut
